wgpu-hal/src/gles/egl.rs

```diff
@@ -758,8 +764,47 @@ impl crate::Instance<super::Api> for Instance {
         let mut inner = self.inner.lock();
 
         match raw_window_handle {
-            Rwh::Xlib(_) => {}
-            Rwh::Xcb(_) => {}
+            #[cfg(not(feature = "emscripten"))]
+            Rwh::Xlib(handle) => {
+                // If Wayland EGLDisplay was created, but we got an X window, recreate the context
+                // https://github.com/gfx-rs/wgpu/issues/2762
+                if inner.wl_display.is_some() || inner.wsi.kind == WindowKind::Wayland {
+                    log::warn!("Re-initializing Gles context due to Xlib window");
+
+                    use std::ops::DerefMut;
+                    let display_attributes = [egl::ATTRIB_NONE];
+                    let display = inner
+                        .egl
+                        .instance
+                        .upcast::<egl::EGL1_5>()
+                        .unwrap()
+                        .get_platform_display(
+                            EGL_PLATFORM_X11_KHR,
+                            handle.display,
+                            &display_attributes,
+                        )
+                        .unwrap();
+
+                    let new_inner = Inner::create(
+                        self.flags,
+                        Arc::clone(&inner.egl.instance),
+                        display,
+                        WindowSystemInterface {
+                            kind: WindowKind::X11,
+                            library: libloading::Library::new("libX11.so").ok().map(Arc::new),
+                        },
+                    )
+                    .map_err(|_| crate::InstanceError)?;
+
+                    let old_inner = std::mem::replace(inner.deref_mut(), new_inner);
+                    inner.wl_display = None;
+                    drop(old_inner);
+                }
+            }
+            Rwh::Xcb(_) => {
+                // TODO: support Xcb context reinit
+                // https://www.khronos.org/registry/EGL/extensions/EXT/EGL_EXT_platform_xcb.txt
```

Comment:
Last time I recall Mesa wasn't advertising this extension? Probably need to check again.

Reply:
I left a TODO just because didn't find any necessity to implement this